curator-framework/src/main/java/org/apache/curator/framework/state/ConnectionStateManager.java

```diff
@@ -319,8 +319,18 @@ else if ( currentConnectionState == ConnectionState.LOST )
         {
             try
             {
-                // give ConnectionState.checkTimeouts() a chance to run, reset ensemble providers, etc.
-                client.getZookeeperClient().getZooKeeper();
+                if ( client.getZookeeperClient().isConnected() )
```

Comment:
right now, the `checkSessionExperiation` is only called `if sessionExpirationPercent > 0`, we should also change that?

Reply:
Maybe - I feel like this is such a hack I wanted to limit it. But, maybe you're right.

Reply:
Check now @chuchao333

Reply:
LGTM now, thanks for the fix.

Reply:
TBH - this makes me nervous. I worry about an infinite loop of resets. I'll think about this some more before I merge.

Reply:
@Randgalt yep, I agree. I would also prefer if we could fix this without firing the unexpected LOST in the first place.

Also TBH, some of the curator code sounds "no obvious defect" instead of "obviously no defect" to me :) I do understand it's complex enough given all the edge cases it has to handle.

Reply:
Much of that backend code needs to be reconsidered. If you're interested it would be a great project. I could help. If, I find the time, I'll rework some of it. For example, the ConnectionState code belongs in the client code, not the framework code. Pulling all that out and re-writing it directly in the client module would simplify things and give us an opportunity to remove some races due to their being two different threads (ZK's and Curator's).